### PR TITLE
[MRG] fix issue with identifiers in the LCA code

### PR DESCRIPTION
**Pull request by ctb**
This fixes some problems with identifiers due to the changes to `SourmashSignature.name` in https://github.com/dib-lab/sourmash/pull/1179. 

This should probably go into 4.0.0...

Brief writeup of my understanding of the issues so far:
* `sig.name` can be empty if no name is given on the signature, but this then causes problems in both `sourmash lca index` and `LCA_Database.insert(...)`
* this also revealed a problem in the reporting on identifiers in `sourmash lca index` - if the identifiers are identical (in this case `''`, but in whatever case) then the number of leftover signatures gets mis-reported.

A particularly simple situation where this arises is in the test `test_index_empty_sketch_name`, where we:

1. sketch some DNA sequences using `sourmash sketch dna`, without providing a name;
2. try to index them with `lca index`;
3. `sig.name` is now empty, which triggers a problem in `LCA_Database.insert(...)` with duplicate identifiers;
4. the number of unused identifiers tracked in `record_no_lineage` is now miscounted as 1 instead of 2, because we used a set.

This PR fixes the following issues:
* `sourmash sketch` and `sourmash compute` no longer create empty signatures from empty files and stdin;
* `sourmash sketch` and `sourmash compute` set `sig.filename` to empty string when filename is `-`;
* in `sourmash lca index`, `record_no_lineage` is now a list instead of set;
* in `sourmash lca index`, `ident` is taken from `sig.filename` if `sig.name` is empty.
* in `LCA_Database`, `ident` defaults to `str(sig)`

TODO:
- XX add tests for new `sketch` filename behavior on stdin (punted to #1348)
- [x] add direct `lca_db.insert(...)` tests
- [x] add test for new `sourmash sig describe` behavior

## Checklist

- [ ] Is it mergeable?
- [ ] `make test` Did it pass the tests?
- [ ] `make coverage` Is the new code covered?
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

### CODE REVIEW

**Comment by ctb** [comment]
I think this is ready, @luizirber and @bluegenes. And I think it should go into 4.0 release, because it is a problem that is triggered by #1179, which is one of the "breaking backwards compatibility" PRs going into 4.0.

File: src/sourmash/command_compute.py [hunk 1]
```diff
@@ -188,12 +191,13 @@ def _compute_individual(args, signatures_factory):
                 add_seq(sigs, record.sequence,
                         args.input_is_protein, args.check_sequence)
 
-            notify('...{} {} sequences', filename, n, end='')
+            if n is not None:       # don't write out signatures if no input
```
Comment (bluegenes) [inline src/sourmash/command_compute.py:196]: is it worth notifying that the input was empty?

File: tests/test_lca.py [hunk 2]
```diff
@@ -69,6 +69,22 @@ def test_api_create_insert_bad_ksize():
         lca_db.insert(ss)
 
 
+def test_api_create_insert_bad_ident():
+    # can we insert a ksize=21 signature into a ksize=31 DB? hopefully not.
```
Comment (bluegenes) [inline tests/test_lca.py]:
```suggestion
    # can we insert a signature with no ident?
```

(I assume this test is meant to pass?) No assertion for this test.

Comment (ctb) [reply tests/test_lca.py]: yep. adding comment in code. thx :)

File: src/sourmash/command_compute.py (hunk 1 above)
Comment (ctb) [reply src/sourmash/command_compute.py:196]: yes, thank you! updated.

**Comment by ctb** [comment]
Fixed! thanks @bluegenes!

**Review by bluegenes — APPROVED** [review]

**Comment by ctb** [comment]
thanks!

**Pull request merged**